core/src/avm2/dynamic_map.rs

```diff
@@ -0,0 +1,215 @@
+                if raw.is_bucket_full(i) {
+                    // SAFETY: We know that this bucket is safe to access because we just checked
+                    // that it is full.
+                    let bucket = raw.bucket(i).as_ref();
+                    if bucket.1.enumerable {
+                        count += 1;
+                        if count >= index {
+                            return Some(i);
+                        }
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    fn raw(&mut self) -> &mut RawTable<(K, DynamicProperty<V>)> {
+        self.values.raw_table_mut()
+    }
+
+    pub fn remove(&mut self, key: &K) -> Option<DynamicProperty<V>> {
+        self.values.remove(key)
+    }
+
+    pub fn next(&mut self, index: usize) -> Option<usize> {
+        // If the public index is zero than this is the first time this is being enumerated,
+        // if index != self.public_index, then we are enumerating twice OR we mutated while enumerating.
+        //
+        // Regardless of the reason, we just need to sync the supplied index to the real index.
+        if self.public_index == 0 || index != self.public_index {
+            if let Some(real) = self.public_to_real_index(index) {
+                self.real_index = real;
+                self.public_index = index + 1;
+                return Some(self.public_index);
+            } else {
+                self.public_index = 0;
+                self.real_index = 0;
+                return None;
+            }
+        }
+
+        let real = self.real_index + 1;
+        let raw = self.raw();
+        let total_buckets = raw.buckets();
+        if !raw.is_empty() && real < total_buckets {
+            for i in real..total_buckets {
+                unsafe {
+                    // SAFETY: It is impossible for i to be greater than the total buckets.
+                    if raw.is_bucket_full(i) {
+                        // SAFETY: We know that this bucket is safe to access because we just checked
+                        // that it is full.
+                        let bucket = raw.bucket(i).as_ref();
+                        if bucket.1.enumerable {
+                            self.real_index = i;
+                            self.public_index += 1;
+                            return Some(self.public_index);
+                        }
+                    }
+                }
+            }
+        }
+        None
+    }
+
+    pub fn pair_at(&mut self, index: usize) -> Option<&(K, DynamicProperty<V>)> {
+        // NOTE: AVM2 object enumeration is one of the weakest parts of an
+        // otherwise well-designed VM. Notably, because of the way they
+        // implemented `hasnext` and `hasnext2`, all enumerants start from ONE.
+        // Hence why we have to `checked_sub` here in case some miscompiled
+        // code doesn't check for the zero index, which is actually a failure
+        // sentinel.
```

Comment:
This talks about a `checked_sub` operation that doesn't appear anywhere; is the comment outdated?

Reply:
I just deleted the whole comment, since we're now more closely based on avm2